### DISCUSSION TITLE
Cannot get `vwt.initializeWithCopy` to copy Arrays properly?

### DISCUSSION THREAD

**Discussion by NSExceptional** [Azoy/Echo#51] [Q&A] [answered]
To recap, I am using Echo to make a JSON object mapper. My library uses the output of `JSONSerialization` to initialize any type you want. So far it works very well, except for cases like this where COW properties are involved. In this case, I have a `Person` type like so, and I am trying to decode a `Person` with one child.

```swift
struct Person {
    /// For testing purposes; this is used as the structure of the "one child"
    static let sally = Person(name: "Sally", age: 5, married: false, kids: [], job: nil)
    
    var name: String
    var age: Int
    var married: Bool
    var kids: [Person]
    var job: String?
}
```

It looks like the array's backing storage is being released out from under it when the `kids` ivar is assigned in `StructMetadata.createInstance()` **edit: but it is not**. To be specific, once the array is assigned to the `box` in `createInstance()`, attempting to print or view the array from the box yields a description of `1 element` but it does not print the element, and trying to access the element yields a failed force-unwrap. I am still able to print the array from the input dictionary:

<img width="216" alt="Screenshot 2021-05-06 at 1 23 43 PM" src="https://user-images.githubusercontent.com/8371943/117347187-5fb0f400-ae6e-11eb-87bb-9edcc2f8f482.png">

This probably would affect other types of COW collections I assume but so far I am only testing Array. The strings I'm using are static as well, so they would never run into this problem. Here are the relevant code segments:

<details>
 <summary>Code</summary>

```swift
// Helper operator intended to cast from one
// pointer type to UnsafeMutableRawPointer
postfix operator ~
postfix func ~<T>(target: T) -> UnsafeMutableRawPointer {
    return unsafeBitCast(target, to: UnsafeMutableRawPointer.self)
}

extension StructMetadata {
    // Entrypoint for creating struct instance from a property map
    func createInstance(props: [String: Any] = [:]) -> Any {
        // Precondition: props must contain a key for every property
        assert({
            let givenKeys = Set(props.keys.map { $0 as String })
            let allKeys = Set(self.descriptor.fields.records.map(\.name))
            return givenKeys == allKeys
        }())

        var box = AnyExistentialContainer(metadata: self)
        for (key, value) in props {
            self.set(value: value, forKey: key, pointer: box.getValueBuffer()~)
        }
        
        // At this point, the struct is populated correctly, except
        // for array properties which fail to unwrap any elements
        return box.toAny
    }

    func set(value: Any, forKey key: String, pointer ptr: UnsafeMutableRawPointer) {
        // These two are absolutely correct
        let offset = self.fieldOffset(for: key)!
        let type = self.fieldType(for: key)!
        ptr.storeBytes(of: value, type: type, offset: offset)
    }
}

extension UnsafeMutableRawPointer {
    func storeBytes(of value: Any, type: Metadata, offset: Int = 0) {
        // Things get wonky here if I don't use my modified version
        // of this function to avoid double-boxed containers from #49
        var box = container(for: value)
        type.vwt.initializeWithCopy((self + offset), box.projectValue()~)
        // What I was doing before, which also didn't work for arrays
        // (self + offset).copyMemory(from: box.projectValue(), byteCount: type.vwt.size)
    }
}

// MARK: Populating AnyExistentialContainer
extension AnyExistentialContainer {
    var toAny: Any {
        return unsafeBitCast(self, to: Any.self)
    }
    
    /// Calls into `projectValue()` but will allocate a box
    /// first if needed for types that are not inline
    mutating func getValueBuffer() -> UnsafeMutableRawPointer {
        // Allocate a box if needed and return it
        if !self.metadata.vwt.flags.isValueInline && self.data.0 == 0 {
            return self.metadata.allocateBoxForExistential(in: &self)~
        }
        
        // We don't need a box or already have one
        return self.projectValue()~
    }
}
```

</details>

What could be going on here?

Edit: after updating to your latest commit, I now get `EXC_BAD_ACCESS` when trying to print the fully initialized instance, but the debugger's variable view shows pretty much the same output as my lldb screenshot above. Also, I am able to print the array if I manually dereference the ivar? O_o

<img width="989" alt="Screenshot 2021-05-06 at 3 58 01 PM" src="https://user-images.githubusercontent.com/8371943/117364718-e0c6b600-ae83-11eb-9f4a-44c70011d2b3.png">

**Comment by Azoy** [comment]
I'm very confused because I'm using your code above and creating my own person instances with kids, and I'm able to see them completely (nothing is wrong on my end).

**Reply by NSExceptional** [reply]
I think I just realized what the issue is. It's funny how revelations like this come to you at the strangest times later…

When I created the array I'm trying to assign to `kids`, it was created as `[Any]`. The property I'm assigning it to is of type `[Person]`.

A new problem arises: how do I create an array at runtime with a specific type only known at runtime? Even if I could create it, how do I populate it??? 😫

This seems like it could be a problem for any generic type, but it will be much harder to solve for collections I think. I bet the runtime has a function for this though… Let me guess, one of the `dynamicCast` functions? Yeah, `swift_dynamicCast` looks like what I need!

Final edit: tests passing!

**Reply by Azoy** [reply]
I wouldn't suggest using `swift_dynamicCast` yourself, I would use something like this:
```swift
for (key, value) in props {
  func unboxValue<T>(_: T.Type) {
    let newValue = value as! T
    self.set(value: newValue, forKey: key, pointer: box.getValueBuffer())
  }

  let fieldType = self.type(of: ...)
  _openExistential(type, do: unboxValue(_:))
}
```
In essence, you're doing the same thing, but in this case you're letting Swift do it for you (which is what I recommend 100x times over). This gets the field's actual type and enables you to cast the value you receive to the actual stored type.

**Reply by Azoy** [reply]
There's another check you can do to only perform this cast when the type you receive isn't the stored property type:
```swift
let record = ...
let fieldType = self.type(of: record.mangledTypeName)!

if container(for: value).type == fieldType {
  // No cast necessary!
  self.set(value: value, forKey: key, pointer: box.getValueBuffer())
  continue
}

// Ok, now we need to open the type and cast. Use what I wrote above here.
```

**Reply by NSExceptional** [reply]
I should clarify, I'm not just using the cast everywhere, only right after I create an array as `[Any]` during JSON decoding—not when I'm setting values. My `createInstance` functions expect everything to be the exact correct types already.

I am using the concrete type of the field as well when I do the cast, so it should be functionally equivalent to the code you suggested with `_openExistential`, right?

I didn't know about `_openExistential` though, that's really cool!

**Reply by Azoy** [reply]
Sure, it might be functionally equivalent, but the issue is that once you step outside of Swift (calling this C function), you remove all optimization that the optimizer can do on the piece of code. You can tweak my code to be able to use it in your case and still have it work as well as get the optimizations. I highly suggest not using `swift_dynamicCast` yourself.

**Reply by NSExceptional** [reply]
Ah, I see. I will make use of it then. Thank you :)

**Reply by NSExceptional** [reply]
I'm now running into a weird issue where the cast succeeds, but the underlying type of the array element doesn't match up… In particular, I'm attempting to cast `NSArray<NSDictionary *>` to `[Post]` where `Post` is a `class` type, and the cast is succeeding, **but I expect it to fail**. As a result it crashes elsewhere when I begin to enumerate the array when the element type doesn't match the type of the array.

Do you know why this would happen? I thought the casting checks the element types

<img width="982" alt="Screenshot 2021-05-12 at 1 40 44 PM" src="https://user-images.githubusercontent.com/8371943/118027568-b2375800-b327-11eb-9942-cc6cb0d2af7f.png">

**Reply by Azoy** [reply]
The same error occurs even without the reflection bits:
```swift
class Person { ... }

let x: NSDictionary = ["name": "Alex", "age": "20"]
let y: NSArray = [x]

// error: Expected Person but found __NSDictionaryI
let z = y as! [Person]
```

I would simply check types before casting to ensure that types are bridgable. I really don't know the specifics of how you're getting `NSArray`s and such, so I don't know the best solution in your case.

**Reply by NSExceptional** [reply]
Oh, weird. I'll do that, thanks.

Switching gears one more time: can I use `_openExistential` to, say, invoke the initializer for an `ExpressibleByStringLiteral` type? Or any other type with an associated type

This doesn't appear to compile:

```swift
class Pet: ExpressibleByStringLiteral {
    typealias StringLiteralType = String
    var name: String
    
    required init(stringLiteral: StringLiteralType) {
        self.name = stringLiteral
    }
}

func makeFromStringLiteral<T: ExpressibleByStringLiteral>(_: T.Type) -> T where T.StringLiteralType == String {
    return T.init(stringLiteral: "fido")
}

let p: Any = _openExistential(Pet.self, makeFromStringLiteral(_:))
```

I'd like to be able to construct an enum instance by calling into `init(rawValue:)`, and ideally a generic way to do it for any raw value type without hard-coding checks for several different types

**Reply by Azoy** [reply]
Could you not just do:
```swift
makeFromStringLiteral(Pet.self)
```
? The way you're using `_openExistential` does not work/doesn't make sense because there is no existential to open.

**Reply by NSExceptional** [reply]
Replace `Pet.self` with a variable of type `Any.Type`, my bad:

```swift
let type: Any.Type = metadata.type
makeFromStringLiteral(type) // nope
```

Another example:

```swift
enum Colors: String {
    case red = "red"
    case blue = "blue"
    case green = "green"
}

let rawValue: Any = ...

func createColor<T: RawRepresentable>(_: T.Type) -> T {
    let value = rawValue as! T.RawValue
    return T.init(rawValue: value)!
}

let type: Any.Type = Colors.self
let i = _openExistential(type, do: createColor(_:))
```

This yields `Local function 'createColor' requires that 'Any' conform to 'RawRepresentable'`

**Reply by Azoy** [reply]
The issue here is that we can't reference those protocols, `associated type and Self requirement!!!!`, but soon we will be able to make existentials with those protocols and the compiler will happily accept what you're trying to do.
```swift
let x = metadata.type as! ExpressibleByStringLiteral.Type
let y = _openExistential(x, do: makeFromStringLiteral(_:))
```

**Reply by NSExceptional** [reply]
Ah, gotcha. I will just work around this until existentials are opened up then. Thanks :)

**Comment by NSExceptional** [comment, marked as answer]
When I created the array I'm trying to assign to `kids`, it was created as `[Any]`. The property I'm assigning it to is of type `[Person]`.

Solution: use `swift_dynamicCast` to perform the conversion :)

Edit: see Azoy's replies above; `swift_dynamicCast` may not be as good a choice as `_openExistential`
